Parameter for generators to define their JDK version

Goldorion opened the pull request:
Since Minecraft 1.17, Minecraft uses Java 16. However, the internal MCreator's JDK is still Java 11, so my generators can't be updated to the newest version as long as Forge doesn't support J16. In addition, after updating the internal JDK for J16, several generators will be unusable. This PR fixes this problem. It adds a new (optional) parameter for custom generators (defined inside the generator.yaml file) allowing them to choose the JDK they need to work. Then, MCreator will check if the user has this JDK downloaded in the new `jdks` folder (of `.mcreator`). If the JDK is not in this folder, MCreator will automatically try to download the AdoptJDK's version for the computer.

I tested this feature with the current Forge 1.16.5 generator and a test generator for Fabric 1.17 and both could launch Minecraft without problems.

Example in this commit: https://github.com/MCreator/MCreator/pull/1454/commits/5db292b1acc3653942121a75ca628929b919af5a

KlemenDEV commented:
@Goldorion indeed loom requires J16 while we are still at J11 and toolchains can't fix this for plugins themselves. I think you can go ahead with this feature but make this optional so by default, bundled JDK is still used so only when this is really needed by some plugin, it is used.

Alternatively, we can wait for forge gradle to support J16 as I will move to it as soon as this is possible and with J16

Goldorion commented:
> 
> 
> @Goldorion indeed loom requires J16 while we are still at J11 and toolchains can't fix this for plugins themselves. I think you can go ahead with this feature but make this optional so by default, bundled JDK is still used so only when this is really needed by some plugin, it is used.
> 
> Alternatively, we can wait for forge gradle to support J16 as I will move to it as soon as this is possible and with J16

I re-added the code to use the bundled JDK, so everything is ready for review.

KlemenDEV commented on src/main/java/net/mcreator/ui/dialogs/workspace/WorkspaceGeneratorSetupDialog.java:
```diff
@@ -46,6 +47,18 @@ public static void runSetup(final MCreator m, boolean showWebsite) {
 		AtomicBoolean setupOk = new AtomicBoolean(true);
 
 		Thread t = new Thread(() -> {
+			ProgressDialog.ProgressUnit p0 = new ProgressDialog.ProgressUnit(
```
This step should only happen if generator specifies jdk

KlemenDEV commented on src/main/java/net/mcreator/gradle/GradleUtils.java:
```diff
@@ -80,6 +82,12 @@ public static String getJavaHome() {
 		}
 
+		// if the generator has a JDK, we set JAVA_HOME to generator's JDK
+		if (UserFolderManager.getSpecificJDK(genConfig.getJDKVersion() + "/bin/javac.exe").isFile() || UserFolderManager
```
This step should be skipped if the generator configuration does not specify jdk. Also refactor genConfig to full name

KlemenDEV commented on src/main/java/net/mcreator/ui/action/impl/workspace/DownloadJDK.java:
```diff
@@ -0,0 +1,68 @@
+import java.io.IOException;
+import java.net.URL;
+
+public class DownloadJDK {
```
A better name would be eg. JDKDownloader, even better if we called this SDK instead of JDK I think

KlemenDEV commented on src/main/java/net/mcreator/gradle/GradleUtils.java:
```diff
@@ -80,6 +82,13 @@ public static String getJavaHome() {
 		}
 
+		// if the generator has a JDK, we set JAVA_HOME to generator's JDK
+		if (generatorConfiguration.getJDKVersion() != null && (UserFolderManager.getSpecificJDK(generatorConfiguration
```
You should do OS specific checks here

KlemenDEV commented on src/main/java/net/mcreator/ui/action/impl/workspace/SDKDownloader.java:
```diff
@@ -0,0 +1,64 @@
+							UserFolderManager.getSpecificJDK(jdkVersion + ".zip"));
+
+					LOG.info("Unzipping the JDK");
+					ZipIO.unzip(UserFolderManager.getSpecificJDK(jdkVersion + ".zip").getAbsolutePath(),
```
Is macOS version also zip? Make sure to test this on all 3 supported platforms. I suggest setting up VM to test this out. I suspect there may be problems, at least on macOS as folder structure of macOS is quite different

KlemenDEV commented on src/main/java/net/mcreator/io/OS.java:
```diff
@@ -36,6 +36,15 @@ public static int getOS() {
 		return WINDOWS;
 	}
 
+	public static String getOSName() {
```
Use existing getOS here so you don't re-do the os name matching logic again here

Goldorion commented:
> I suggest setting up VM to test this out. 
@KlemenDEV I can't test on MacOS even with a VM because Apple doesn't allow to use another computer than their own Mac. If we want to test on MacOS, @Max094Reikeb will have to do it. If he doesn't have a problem, Linux will also work as both don't have file extensions and use a `.tar.gz` archive file.

KlemenDEV commented:
> > I suggest setting up VM to test this out.
> 
> @KlemenDEV I can't test on MacOS even with a VM because Apple doesn't allow to use another computer than their own Mac. If we want to test on MacOS, @Max094Reikeb will have to do it. If he doesn't have a problem, Linux will also work as both don't have file extensions and use a `.tar.gz` archive file.

Linux has different folder structure for JDK than macOS so both platforms need testing

KlemenDEV commented on src/main/java/net/mcreator/ui/action/impl/workspace/SDKDownloader.java:
```diff
@@ -0,0 +1,83 @@
+			if (!UserFolderManager.getSpecificJDK(jdkVersion + "/").exists()) {
+				LOG.info("Downloading JDK: " + jdkVersion);
+				try {
+					if (OS.getOS() == OS.WINDOWS) {
```
Oh I did not mean that, I just meant that use getOS() ... in OS class when returning name. It was ok and better before, I just wanted you did not compare OS name by string again as this logic is defined in getOS. Sorry for the confusion

KlemenDEV replied:
Changed to draft so this gets resolved first

KlemenDEV commented on src/main/java/net/mcreator/ui/dialogs/workspace/WorkspaceGeneratorSetupDialog.java:
```diff
@@ -46,6 +47,20 @@ public static void runSetup(final MCreator m, boolean showWebsite) {
+				if (SDKDownloader.downloadJDK(m.getGeneratorConfiguration())) {
+					p0.ok();
+				} else {
+					p0.err();
```
I think that if generator downloading fails, you should stop setup here and report error as without JDK, no step further will work

KlemenDEV commented on src/main/java/net/mcreator/io/OS.java:
```diff
@@ -36,6 +36,14 @@ public static int getOS() {
+		if (getOS() == LINUX)
+			return "linux";
+		return "windows";
+	}
```
You have re-added the method, but are not using it

KlemenDEV commented:
@Goldorion could you also send me a plugin zip that uses this feature so I can try and test it?

Goldorion commented:
[generator-1.16.5.zip](https://github.com/MCreator/MCreator/files/6726321/generator-1.16.5.zip)

KlemenDEV commented on src/main/java/net/mcreator/ui/action/impl/workspace/SDKDownloader.java:
```diff
@@ -0,0 +1,65 @@
+			else
+				fileExtension = ".tar.gz";
+
+			FileUtils.copyURLToFile(
```
I am not entirely sure how this can work, this downloads zip or tar and you then specify folder for the java home. You will need to extract the archive in the appropriate folder for this to work

Goldorion replied:
I tried several things for ProgressMonitorInputStream, but the dialog was never visible.

KlemenDEV replied:
I presume the UI then blocks when you download which is not ok. Also you need to add task to unzip the downloaded files

KlemenDEV reviewed (changes requested):
Also consider using https://docs.oracle.com/javase/7/docs/api/javax/swing/ProgressMonitorInputStream.html to show download progress

Goldorion commented:
@KlemenDEV I tried several things for ProgressMonitorInputStream, but the dialog was never visible.

KlemenDEV commented:
Do you plan on fixing this?

Goldorion commented:
> 
> 
> Do you plan on fixing this?

I'm not able to use ProgressMonitorInputStream to have the progress bar.

KlemenDEV commented:
> > Do you plan on fixing this?
> 
> I'm not able to use ProgressMonitorInputStream to have the progress bar, but everything else works.

You are not extracting the zip, only downloading it, this can not work

Goldorion commented:
> 
> 
> > > Do you plan on fixing this?
> > 
> > 
> > I'm not able to use ProgressMonitorInputStream to have the progress bar, but everything else works.
> 
> You are not extracting the zip, only downloading it, this can not work

I added it when I created the PR. Weird :/ I'll add it again.

KlemenDEV commented:
Also on nix systems you will need to untar, not unzip, I think

Goldorion commented:
The unzipping code for Windows works fine. However, concerning Mac and Linux I tried something, but @Max094Reikeb or another Mac/Linux user will need to test.

Goldorion commented:
> 
> 
> The unzipping code for Windows works fine. However, concerning Mac and Linux I tried something, but @Max094Reikeb or another Mac/Linux user will need to test.

If my current way doesn't work, I will try to install and use Apache Commons Compress as they include a method to untar a file. http://commons.apache.org/proper/commons-compress/

KlemenDEV commented:
Using unzip for tar will not work. I also suggest you install ubuntu VM for testing as you will often need this likely

Goldorion commented:
> Using unzip for tar will not work. I also suggest you install ubuntu VM for testing as you will often need this likely

I'll see if I can install one without breaking my laptop or install a malware. xD

Goldorion commented:
I installed a VM with Ubuntu 20.04 LTS (without breaking everything on my computer!) and I fixed the extracting method for Mac and Linux. As I said before, I added Apache Commons Compress 1.21 to the project, so we can easily extract `.tar.gz` archives. The library also supports several file types, so it may be useful in the future. However, I'm still not able to have the progress bar as you wanted, but I don't think it's useful because no other place in the software has a progress bar (even when downloading a new version of Forge/Fabric/Spigot).

Goldorion commented:
I thought to a small thing. As each generator will be able to have its own JDK version and it's an important thing for the development, I thought to add a line saying the JDK version used by the generator in the generator selector dialog. Should I add this?

KlemenDEV commented:
It would be a nice addition indeed. I will take a look for progress bar if I get some extra time too

Goldorion commented:
When using the default (built-in) JDK: 
![image](https://user-images.githubusercontent.com/38427877/125707687-749f243a-626a-471a-a2ff-c2e5e74396b5.png)

When using a custom JDK defined by the generator:
![image](https://user-images.githubusercontent.com/38427877/125707735-247126d7-6914-4295-a633-256d1f5cf949.png)

KlemenDEV reviewed (changes requested):
I have tested this on Ubuntu and get this error:

> Caused by: org.gradle.api.GradleException: Unable to determine version for JDK located at /home/pylo/.mcreator/sdks/jdk-11.0.11+9. Reason: A problem occurred starting process 'command '/home/pylo/.mcreator/sdks/jdk-11.0.11+9/bin/java''

So something needs to be changed it seems for Linux, maybe for macOS too to work properly.

KlemenDEV commented on src/main/java/net/mcreator/ui/action/impl/workspace/SDKManager.java:
```diff
@@ -67,6 +71,13 @@ public static void downloadSDK(GeneratorConfiguration generatorConfiguration,
+
+			if (OS.getOS() != OS.WINDOWS) {
+				Set<PosixFilePermission> perms = new HashSet<>();
+				perms.add(PosixFilePermission.OWNER_READ);
```
You should set permissions based on the settings from the tar file as I showed on the page I linked, do not hardcode permissions, also your permissions as you set them will not help anyway as you need to handle execute bits too.
